File: internal/tokens/stream.go
// Package tokens holds a flat, mutable, index-addressable token stream. This is
// the ECS/PHP-CS-Fixer working model: fixers read tokens by index and insert,
// replace or remove them in place, then Render rebuilds the source.
package tokens

import (
	"strings"

	"ecs-go/internal/token"
)

type Stream struct {
	toks []token.Token
}

func New(toks []token.Token) *Stream {
	return &Stream{toks: toks}
}

func (s *Stream) Len() int { return len(s.toks) }

func (s *Stream) At(i int) token.Token { return s.toks[i] }

func (s *Stream) Set(i int, t token.Token) { s.toks[i] = t }

// SetValue replaces only the Value of the token at i, keeping its kind.
func (s *Stream) SetValue(i int, v string) { s.toks[i].Value = v }

func (s *Stream) RemoveAt(i int) {
	s.toks = append(s.toks[:i], s.toks[i+1:]...)
}

func (s *Stream) InsertAt(i int, t token.Token) {
	s.toks = append(s.toks, token.Token{})
	copy(s.toks[i+1:], s.toks[i:])
	s.toks[i] = t
}

// MatchForward returns the index of the delimiter matching the opener at i
// ("(", "{" or "["), or -1 if there is none. Strings and comments are single
// tokens, so scanning punctuation is safe.
func (s *Stream) MatchForward(i int) int {
	if i < 0 || i >= len(s.toks) || s.toks[i].Kind != token.Punct {
		return -1
	}
	open := s.toks[i].Value
	var closer string
	switch open {
	case "(":
		closer = ")"
	case "{":
		closer = "}"
	case "[":
		closer = "]"
	default:
		return -1
	}
	depth := 0
	for j := i; j < len(s.toks); j++ {
		if s.toks[j].Kind != token.Punct {
			continue
		}
		switch s.toks[j].Value {
		case open:
			depth++
		case closer:
			depth--
			if depth == 0 {
				return j
			}
		}
	}
	return -1
}

// ReplaceRange swaps tokens [start, end] (inclusive) for repl.
func (s *Stream) ReplaceRange(start, end int, repl []token.Token) {
	tail := append([]token.Token(nil), s.toks[end+1:]...)
	s.toks = append(s.toks[:start], repl...)
	s.toks = append(s.toks, tail...)
}

// Render concatenates every token value back into source.
func (s *Stream) Render() string {
	var b strings.Builder
	for _, t := range s.toks {
		b.WriteString(t.Value)
	}
	return b.String()
}

// Tokens returns the underlying slice (read-only use).
func (s *Stream) Tokens() []token.Token { return s.toks }

File: rust/src/stream.rs
// A flat, mutable, index-addressable token stream, ported from ecs-go's
// internal/tokens. Fixers read tokens by index and insert, replace or remove
// them in place; render rebuilds the source.

use crate::token::Token;

pub struct Stream {
    pub toks: Vec<Token>,
}

impl Stream {
    pub fn new(toks: Vec<Token>) -> Self {
        Stream { toks }
    }

    pub fn len(&self) -> usize {
        self.toks.len()
    }

    pub fn at(&self, i: usize) -> &Token {
        &self.toks[i]
    }

    pub fn set_value(&mut self, i: usize, v: &[u8]) {
        self.toks[i].value = v.to_vec();
    }

    pub fn remove_at(&mut self, i: usize) {
        self.toks.remove(i);
    }

    pub fn insert_at(&mut self, i: usize, t: Token) {
        self.toks.insert(i, t);
    }

    pub fn render(&self) -> Vec<u8> {
        let cap = self.toks.iter().map(|t| t.value.len()).sum();
        let mut out = Vec::with_capacity(cap);
        for t in &self.toks {
            out.extend_from_slice(&t.value);
        }
        out
    }
}
